File: solutions/165-compare-version-numbers/compare-version-numbers.java
class Solution {
    public Pair<Integer, Integer> getNextChunk(String version, int n, int p) {
        //If the pointer is set to the end of the string, return 0
        if (p > n - 1) {
            return new Pair(0, p);
        }

        // Find the end of chunk
        int i, pEnd = p;
        while (pEnd < n && version.charAt(pEnd) != '.') {
            ++pEnd;
        }

        // Retrieve the chunk
        if (pEnd != n - 1) {
            i = Integer.parseInt(version.substring(p, pEnd));
        } else {
            i = Integer.parseInt(version.substring(p, n));
        }

        // Find the beginning of the next chunk
        p = pEnd + 1;

        return new Pair(i, p);
    }

    public int compareVersion(String version1, String version2) {
        int p1 = 0, p2 = 0;
        int n1 = version1.length(), n2 = version2.length();

        // Compare versions
        int i1, i2;
        Pair<Integer, Integer> pair;
        while (p1 < n1 || p2 < n2) {
            pair = getNextChunk(version1, n1, p1);
            i1 = pair.getKey();
            p1 = pair.getValue();

            pair = getNextChunk(version2, n2, p2);
            i2 = pair.getKey();
            p2 = pair.getValue();
            if (i1 != i2) {
                return i1 > i2 ? 1 : -1;
            }
        }

        // The versions are equal
        return 0;
    }
}

File: solutions/165-compare-version-numbers/compare-version-numbers.js
var compareVersion = function (version1, version2) {
    var getNextChunk = function (version, n, p) {
        if (p > n - 1) {
            return [0, p];
        }
        let pEnd = p;
        while (pEnd < n && version[pEnd] != ".") {
            pEnd++;
        }
        let i = parseInt(version.substring(p, pEnd != n - 1 ? pEnd : n));
        p = pEnd + 1;
        return [i, p];
    };

    let p1 = 0,
        p2 = 0;
    let n1 = version1.length,
        n2 = version2.length;
    while (p1 < n1 || p2 < n2) {
        let [i1, p1_] = getNextChunk(version1, n1, p1);
        p1 = p1_;
        let [i2, p2_] = getNextChunk(version2, n2, p2);
        p2 = p2_;
        if (i1 != i2) {
            return i1 > i2 ? 1 : -1;
        }
    }
    return 0;
};